File: dotnet/samples/DocumentationExamples/UsingTheKernel.cs
// Copyright (c) Microsoft. All rights reserved.

using System.Threading.Tasks;
// <NecessaryPackages>
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Logging;
using Microsoft.SemanticKernel;
using Microsoft.SemanticKernel.Plugins.Core;
// </NecessaryPackages>
using Xunit;
using Xunit.Abstractions;

namespace Examples;

/// <summary>
/// This example demonstrates how to interact with the kernel as described at
/// https://learn.microsoft.com/semantic-kernel/agents/kernel
/// </summary>
public class UsingTheKernel : BaseTest
{
    [Fact]
    public async Task RunAsync()
    {
        WriteLine("======== Kernel ========");

        string? endpoint = TestConfiguration.AzureOpenAI.Endpoint;
        string? modelId = TestConfiguration.AzureOpenAI.ChatModelId;
        string? apiKey = TestConfiguration.AzureOpenAI.ApiKey;

        if (endpoint is null || modelId is null || apiKey is null)
        {
            WriteLine("Azure OpenAI credentials not found. Skipping example.");

            return;
        }

        // Create a kernel with a logger and Azure OpenAI chat completion service
        // <KernelCreation>
        var builder = Kernel.CreateBuilder()
                            .AddAzureOpenAIChatCompletion(modelId, endpoint, apiKey);
        builder.Services.AddLogging(c => c.AddDebug().SetMinimumLevel(LogLevel.Trace));
        builder.Plugins.AddFromType<TimePlugin>();
        builder.Plugins.AddFromPromptDirectory("./../../../Plugins/WriterPlugin");
        Kernel kernel = builder.Build();
        // </KernelCreation>

        // Get the current time
        // <InvokeUtcNow>
        var currentTime = await kernel.InvokeAsync("TimePlugin", "UtcNow");
        WriteLine(currentTime);
        // </InvokeUtcNow>

        // Write a poem with the WriterPlugin.ShortPoem function using the current time as input
        // <InvokeShortPoem>
        var poemResult = await kernel.InvokeAsync("WriterPlugin", "ShortPoem", new()
        {
            { "input", currentTime }
        });
        WriteLine(poemResult);
        // </InvokeShortPoem>
    }

    public UsingTheKernel(ITestOutputHelper output) : base(output)
    {
    }
}

File: python/samples/learn_resources/using_the_kernel.py
# Copyright (c) Microsoft. All rights reserved.

# <NecessaryPackages>
import asyncio
import os

from samples.sk_service_configurator import add_service
from semantic_kernel import Kernel

# </NecessaryPackages>


async def main():
    # Initialize the kernel
    # <KernelCreation>
    kernel = Kernel()
    # Add the service to the kernel
    # use_chat: True to use chat completion, False to use text completion
    kernel = add_service(kernel, use_chat=True)
    # </KernelCreation>

    # <InvokeUtcNow>
    # Import the TimePlugin and add it to the kernel
    from semantic_kernel.core_plugins import TimePlugin

    time = kernel.add_plugin(TimePlugin(), "TimePlugin")

    # Invoke the Today function
    current_time = await kernel.invoke(time["today"])
    print(f"The current date is: {current_time}\n")
    # </InvokeUtcNow>

    # <InvokeShortPoem>
    # Import the WriterPlugin from the plugins directory.
    script_directory = os.path.dirname(__file__)
    plugins_directory = os.path.join(script_directory, "plugins")
    kernel.add_plugin(parent_directory=plugins_directory, plugin_name="WriterPlugin")
    # Run the short poem function with the Kernel Argument
    poem_result = await kernel.invoke(
        function_name="ShortPoem", plugin_name="WriterPlugin", input=str(current_time)
    )
    print(f"The poem result:\n\n{poem_result}")
    # </InvokeShortPoem>


# Run the main function
if __name__ == "__main__":
    asyncio.run(main())
